### PR TITLE
Updated the authentication configuration to support credentials

### LINKED ISSUE #42: Add configuration for autentication
I would like to configure different clients with authentication. 

```yaml
httplug:
  clients:
    guzzle5:
      factory: 'httplug.factory.guzzle5'
      plugins: ['httplug.plugin.logger']
    trans_logger:
      factory: 'httplug.factory.guzzle5'
      authentication: 
        type: 'basic'
        username: 'foo'
        password: 'bar'
    microservice2:
      factory: 'httplug.factory.guzzle5'
      plugins: ['httplug.plugin.logger', 'my_cache_plugin']
      authentication: 
        type: 'wsse'
        username: 'baz'
        password: 'biz'
```

What do you think? Is this something we should support? Do you like the config example?

### PR DESCRIPTION
**Pull request by Nyholm**
This PR fixes #42 

```yaml
// config.yml
httplug:
    plugins:
        authentication:
            my_basic:
                type: 'basic'
                username: 'foo'
                password: 'bar'
            my_wsse:
                type: 'wsse'
                username: 'foo'
                password: 'bar'
            my_brearer:
                type: 'bearer'
                token: 'foo'

    clients:
        acme:
            factory: 'httplug.factory.guzzle6'
            plugins: ['httplug.plugin.authentication.my_wsse']
```

This PR will break BC because change of the configuration in `httplug.plugins.authentication`.

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -120,14 +120,18 @@ protected function configurePlugins(ArrayNodeDefinition $root)
...
+                        ->prototype('array')
+                            ->children()
+                                ->enumNode('type')
+                                    ->values(['basic', 'bearer', 'wsse'])
+                                    ->isRequired()
+                                    ->cannotBeEmpty()
+                                ->end()
+                                ->scalarNode('username')->end()
+                                ->scalarNode('password')->end()
+                                ->scalarNode('token')->end()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: we could go all fancy with custom validation to have the right required parameters based on the type. but we don't need that now.

i wonder if we should add type `service` that accepts a scalarNode `authentication_service` to keep supporting custom authentication services?

File: DependencyInjection/HttplugExtension.php [hunk 2]
```diff
@@ -162,4 +168,53 @@ private function configurePluginByName($name, Definition $definition, array $con
...
+     */
+    private function configureAuthentication(ContainerBuilder $container, array $config)
+    {
+        foreach ($config as $name => $values) {
+            $authServiceKey = sprintf('httplug.plugin.authentication.%s.auth', $name);
+            if ($values['type'] === 'bearer') {
+                $this->validateAuthenticationConfiguration($values, 'bearer', $name, ['token']);
+                $container->register($authServiceKey, Bearer::class)
+                    ->addArgument($values['token']);
+            } elseif ($values['type'] === 'basic') {
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: i would use a switch statement for this.

File: README.md [hunk 3]
```diff
@@ -134,6 +134,31 @@ httpug:
...
+                type: 'basic'
+                username: 'foo'
+                password: 'bar'
+            my_wsse:
+                type: 'wsse'
+                username: 'foo'
+                password: 'bar'
+            my_brearer:
+                type: 'bearer'
+                token: 'foo'
```
Comment (dbu) [inline README.md]: can we use something else than "foo" here to make it visually different? maybe 'hash' ?

File: DependencyInjection/HttplugExtension.php [hunk 4]
```diff
@@ -162,4 +168,53 @@ private function configurePluginByName($name, Definition $definition, array $con
...
+
+    /**
+     * Throw a Logic error with a descriptive error if some configuration is missing.
+     *
+     * @param array  $config
+     * @param string $authType
+     * @param string $type
+     * @param array  $names
+     */
+    private function validateAuthenticationConfiguration(array $config, $type, $authType, array $names)
```
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: ok, as we need the validation anyways, could we move it to the Configuration class? an example of how this looks would be https://github.com/doctrine/DoctrinePHPCRBundle/blob/master/DependencyInjection/Configuration.php#L133-L152

**Comment by dbu** [comment]
perfect, i think this configuration looks as simple and straightforward as it can be!

File: DependencyInjection/HttplugExtension.php (hunk 4 above)
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: Great! Thank you for this example. I will remove the `validateAuthenticationConfiguration` and use something like this instead.

File: DependencyInjection/Configuration.php (hunk 1 above)
Comment (Nyholm) [inline DependencyInjection/Configuration.php]: > i wonder if we should add type service that accepts a scalarNode authentication_service to keep supporting custom authentication services?

Sure!

Comment (sagikazarmark) [inline DependencyInjection/Configuration.php]: Not sure. People can still set up authentication plugins, can't they?

Comment (Nyholm) [inline DependencyInjection/Configuration.php]: If we add type `service` then you could use the authentication plugin just as before. It just require you to do a minor config change. 

Before: 
```yaml
// config.yml
httplug:
    plugins:
        authentication:
             authentication: 'my_authentication_service'
    clients:
        my_client:
            factory: 'httplug.factory.guzzle6'
            plugins: ['httplug.plugin.authentication']
```

After:
```yaml
// config.yml
httplug:
    plugins:
        authentication:
            acme:
                type: 'service'
                service: 'my_authentication_service'
    clients:
        my_client:
            factory: 'httplug.factory.guzzle6'
            plugins: ['httplug.plugin.authentication.acme']
```

**Comment by Nyholm** [comment]
Thank you for the feedback. I've addressed your comments.

File: DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -235,4 +225,62 @@ protected function configurePlugins(ArrayNodeDefinition $root)
...
+                    ->always()
+                    ->then(function ($config) {
+                        switch ($config['type']) {
+                            case 'basic':
+                                if (empty($config['username']) || empty($config['password'])) {
+                                    throw new InvalidConfigurationException('Authentication "basic" requires both "username" and "password".');
+                                }
+                                break;
+                            case 'bearer':
+                                if (empty($config['bearer'])) {
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: s/bearer/token/

File: DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -235,4 +225,62 @@ protected function configurePlugins(ArrayNodeDefinition $root)
...
+            ->useAttributeAsKey('name')
+            ->prototype('array')
+                ->validate()
+                    ->always()
+                    ->then(function ($config) {
+                        switch ($config['type']) {
+                            case 'basic':
+                                if (empty($config['username']) || empty($config['password'])) {
+                                    throw new InvalidConfigurationException('Authentication "basic" requires both "username" and "password".');
+                                }
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: should we also complain about fields that should *not* be set? maybe too strict though.

Comment (Nyholm) [inline DependencyInjection/Configuration.php]: Good question. I believe it is not a required feature but maybe a *nice to have*. 
Maybe we should do something more sophisticated then. Do you know if it is possible to conditionally add nodes?

Like "If $config['type'] == 'basic', add node 'username' and 'password'"

Comment (dbu) [inline DependencyInjection/Configuration.php]: the nodes happen before the configuration comes. so i think we would need to check `if (!empty(...)`

Comment (Nyholm) [inline DependencyInjection/Configuration.php]: > the nodes happen before the configuration comes. so i think we would need to check if (!empty(...)

Sorry, not too sure I understand what you mean.

Comment (dbu) [inline DependencyInjection/Configuration.php]: sorry. what i meant is: the code creating the configuration nodes is executed without knowledge of the concrete configuration. only the callback of the `validate()` can know the actual configuration, so i don't think we can define the nodes conditionally.

Comment (dbu) [inline DependencyInjection/Configuration.php]: created #48 to show how i propose to do this

**Comment by Nyholm** [comment]
I've merged PR #48 and I think this PR is ready for merge into master.

**Comment by sagikazarmark** [comment]
A squash would be a good idea first.

**Comment by Nyholm** [comment]
Of course, Sorry @sagikazarmark. 
They are now squashed.

**Comment by dbu** [comment]
great job, thanks a lot!

**Comment by Nyholm** [comment]
Thank you for the input!

**Pull request merged**